packages/fetchai/connections/prometheus/connection.py

```diff
@@ -0,0 +1,314 @@
+        """
+        Handle messages to prometheus.
+
+        :param envelope: the envelope
+        :return: None
+        """
+        enforce(
+            isinstance(envelope.message, PrometheusMessage),
+            "Message not of type PrometheusMessage",
+        )
+        message, dialogue = self._get_message_and_dialogue(envelope)
+
+        if dialogue is None:
+            self.logger.warning(
+                "Could not create dialogue from message={}".format(message)
+            )
+            return
+
+        if message.performative == PrometheusMessage.Performative.ADD_METRIC:
+
+            if message.title in self.metrics:
+                response_code = 409
+                response_msg = "Metric already exists."
+            else:
+                metric_type = getattr(prometheus_client, message.type, None)
+                if metric_type is None:
+                    response_code = 404
+                    response_msg = (
+                        f"{message.type} is not a recognized prometheus metric."
+                    )
+                else:
+                    self.metrics[message.title] = metric_type(
+                        message.title, message.description
+                    )
+                    response_code = 200
+                    response_msg = (
+                        f"New {message.type} successfully added: {message.title}."
+                    )
+
+        elif message.performative == PrometheusMessage.Performative.UPDATE_METRIC:
+
+            metric = message.title
+            if metric not in self.metrics:
+                response_code = 404
+                response_msg = f"Metric {metric} not found."
+            else:
+                update_func = getattr(self.metrics[metric], message.callable, None)
+                if update_func is None:
+                    response_code = 400
+                    response_msg = f"Update function {message.callable} not found for metric {metric}."
+                else:
+                    # Update the metric
+                    update_func(message.value)
+                    response_code = 200
+                    response_msg = f"Metric {metric} successfully updated."
+
+        msg = dialogue.reply(
+            performative=PrometheusMessage.Performative.RESPONSE,
+            target_message=message,
+            code=response_code,
+            message=response_msg,
+        )
+        context = cast(EnvelopeContext, envelope.context)
+        envelope = Envelope(
+            to=msg.to,
+            sender=msg.sender,
+            protocol_id=msg.protocol_id,
+            message=msg,
+            context=context,
+        )
+        await self._send(envelope)
+
+    async def _send(self, envelope: Envelope) -> None:
+        """Send a message.
+
+        :param envelope: the envelope
+        :return: None
+        """
+        await self.queue.put(envelope)
+
+    async def disconnect(self) -> None:
```

Comment:
what about prometheus_client?

Reply:
Right, as @solarw also pointed out, this library somehow does not include stop server functionality. Hopefully the package he suggested can do this correctly.